hivemind/utils/mpfuture.py

```diff
@@ -4,169 +4,209 @@
 import concurrent.futures._base as base
 import multiprocessing as mp
 import multiprocessing.connection
-import time
-from functools import lru_cache
-from typing import Optional, Tuple, Generic, TypeVar
+import os
+import threading
+from typing import Tuple, Generic, TypeVar, Dict, Optional, Any, Callable
 
-from hivemind.utils.threading import run_in_background
+import torch
```

Comment:
Note on torch: it is indeed weird, but so far we're still not sure how else to implement shared value for py3.7

Options:
- current version (with torch.empty)
- mp.Value or mp.Event - cannot send to other processes (cannot serialize)
- using [multiprocessing.shared_memory](https://docs.python.org/3/library/multiprocessing.shared_memory.html) - incompatible with py3.7 (and thus colab & kaggle kernels)
- using [_posixshmem](https://github.com/dillonalaird/shared_numpy/blob/master/shared_numpy/posixshmem.c) (extra dependency to requirements.txt)
- using mp.Pipe - back to where we started, will need an extra pipe per each future